conjure-go-client/httpclient/body.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2024 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpclient
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// RequestBody is an interface that can be used to set the body of an http.Request.
+// Implementations of this interface should set the fields Body, GetBody, and ContentLength (if known).
+type RequestBody interface {
+	setRequestBody(req *http.Request) error
+}
+
+// requestBodyFunc is a function that sets the body of an http.Request and implements RequestBody.
+type requestBodyFunc func() (length int64, body io.ReadCloser, getBody func() (io.ReadCloser, error), err error)
+
+func (f requestBodyFunc) setRequestBody(req *http.Request) (err error) {
+	req.ContentLength, req.Body, req.GetBody, err = f()
+	return err
+}
+
+// RequestBodyInMemory sets the *http.Request Body field to the provided *bytes.Buffer, *bytes.Reader, or *strings.Reader for upload.
+// The GetBody field is set to a function that returns the same io.ReadCloser.
+func RequestBodyInMemory[T bytes.Buffer | bytes.Reader | strings.Reader](input *T) RequestBody {
```

Comment:
Very open to feedback on naming and caller ergonomics on these, they will be called heavily in consumer code.